docs/source/type_narrowing.rst

```diff
@@ -0,0 +1,324 @@
+  the variable at all. When a ``TypeIs`` function returns ``False``, type checkers can narrow
+  the type of the variable to exclude the ``TypeIs`` type.
+
+This behavior can be seen in the following example::
+
+    from typing import TypeGuard, TypeIs, reveal_type, final
+
+    class Base: ...
+    class Child(Base): ...
+    @final
+    class Unrelated: ...
+
+    def is_base_typeguard(x: object) -> TypeGuard[Base]:
+        return isinstance(x, Base)
+
+    def is_base_typeis(x: object) -> TypeIs[Base]:
+        return isinstance(x, Base)
+
+    def use_typeguard(x: Child | Unrelated) -> None:
+        if is_base_typeguard(x):
+            reveal_type(x)  # Base
+        else:
+            reveal_type(x)  # Child | Unrelated
+
+    def use_typeis(x: Child | Unrelated) -> None:
+        if is_base_typeis(x):
+            reveal_type(x)  # Child
+        else:
+            reveal_type(x)  # Unrelated
+
+
+Safety and soundness
+--------------------
+
+While type narrowing is important for typing real-world Python code, many
+forms of type narrowing are unsafe in the presence of mutability. Type checkers
+attempt to limit type narrowing in a way that minimizes unsafety while remaining
+useful, but not all safety violations can be detected.
+
+Incorrect ``TypeIs`` and ``TypeGuard`` functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Both ``TypeIs`` and ``TypeGuard`` rely on the user writing a function that
+returns whether an object is of a particular type. However, the type checker
+does not validate whether the function actually behaves as expected. If it
+does not, the type checker's narrowing behavior will not match what happens
+at runtime.::
+
+    from typing import TypeIs
+
+    def is_str(x: object) -> TypeIs[str]:
+        return True
+
+    def takes_str_or_int(x: str | int) -> None:
+        if is_str(x):
+            print(x + " is a string")  # runtime error
+
+To avoid this problem, every ``TypeIs`` and ``TypeGuard`` function should be
+carefully reviewed and tested.
+
+Unsound ``TypeGuard`` narrowing
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Unlike ``TypeIs``, ``TypeGuard`` can narrow to a type that is not a subtype of the
+original type. This allows for unsafe behavior with invariant data structures::
+
+    from typing import Any, TypeGuard
+
+    def is_int_list(x: list[Any]) -> TypeGuard[list[int]]:
+        return all(isinstance(i, int) for i in x)
+
+    def maybe_mutate_list(x: list[Any]) -> None:
+        if is_int_list(x):
+            x.append(0)  # OK, x is narrowed to list[int]
+
+    def takes_bool_list(x: list[bool]) -> None:
+        maybe_mutate_list(x)
+        reveal_type(x)  # list[bool]
+        assert all(isinstance(i, bool) for i in x)  # fails at runtime
+
+    takes_bool_list([True, False])
+
+To avoid this problem, use ``TypeIs`` instead of ``TypeGuard`` where possible.
+If you must use ``TypeGuard``, avoid narrowing across incompatible types.
```

Comment:
Possibly some better (and certainly more general) advice is to try (though it's not always possible) to use covariant, immutable types for parameter annotations wherever possible. If you do so, there will be many more cases where you'll be able to use `TypeIs`, and so the problem here just arises much less in the first place

Reply:
Maybe this?
```suggestion
To avoid this problem, use ``TypeIs`` instead of ``TypeGuard`` where possible.
If you must use ``TypeGuard``, avoid narrowing across incompatible types.
Prefer using covariant, immutable types in parameter annotations (e.g.,
``Sequence`` or ``Iterable`` instead of ``list``. If you do this, it is more likely
that you'll be able to use ``TypeIs`` to implement your type narrowing functions.
```

Reply:
That looks good to me! Though you need a closing parenthesis after "instead of `list`"

Reply:
It looks like you marked this as resolved but didn't push the change, was that deliberate? :)